botocore/docs/utils.py

```diff
@@ -135,4 +162,18 @@ def traverse_and_document_shape(documenter, section, shape, history,
                     include=include, exclude=exclude,
                     is_top_level_param=is_top_level_param,
                     is_required=is_required)
+        if is_top_level_param:
```

Comment:
I know that this was extracted out before because of the deep inheritance structure, but given the most recent changes I think you can make a much stronger justification for bringing this into a base class as we're starting to copy/paste the `__init__`  of the various documenter classes.  What if we make a base class that contains only this function for shape traversal and have the other base classes share this.  I also suggest this because it's a bit odd to reach into an instance and call one of its collaborators (event_emitter).